File: objc/JawalySMSService.m
#import "JawalySMSService.h"

@interface JawalySMSService ()
@property (nonatomic, strong) NSString *baseURL;
@property (nonatomic, strong) NSString *appID;
@property (nonatomic, strong) NSString *appSecret;
@property (nonatomic, strong) NSURLSession *session;
@end

@implementation JawalySMSService

- (instancetype)initWithAppID:(NSString *)appID appSecret:(NSString *)appSecret {
    self = [super init];
    if (self) {
        _baseURL = @"https://api-sms.4jawaly.com/api/v1/";
        _appID = appID;
        _appSecret = appSecret;
        _session = [NSURLSession sharedSession];
    }
    return self;
}

- (NSString *)authHeader {
    NSString *credentials = [NSString stringWithFormat:@"%@:%@", self.appID, self.appSecret];
    NSData *credentialsData = [credentials dataUsingEncoding:NSUTF8StringEncoding];
    return [NSString stringWithFormat:@"Basic %@", [credentialsData base64EncodedStringWithOptions:0]];
}

- (void)sendSMSWithMessage:(NSString *)message
                  numbers:(NSArray<NSString *> *)numbers
                  sender:(NSString *)sender
              completion:(void (^)(NSString * _Nullable, NSError * _Nullable))completion {
    
    NSString *endpoint = [self.baseURL stringByAppendingString:@"account/area/sms/send"];
    NSURL *url = [NSURL URLWithString:endpoint];
    
    if (!url) {
        NSError *error = [NSError errorWithDomain:@"Invalid URL" code:-1 userInfo:nil];
        completion(nil, error);
        return;
    }
    
    NSDictionary *messageData = @{
        @"messages": @[
            @{
                @"text": message,
                @"numbers": numbers,
                @"sender": sender
            }
        ]
    };
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"POST";
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Accept"];
    [request setValue:[self authHeader] forHTTPHeaderField:@"Authorization"];
    request.HTTPBody = [NSJSONSerialization dataWithJSONObject:messageData options:0 error:nil];
    
    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            completion(nil, error);
            return;
        }
        
        if (!data) {
            NSError *noDataError = [NSError errorWithDomain:@"No data received" code:-1 userInfo:nil];
            completion(nil, noDataError);
            return;
        }
        
        NSError *jsonError;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
        
        if (jsonError) {
            completion(nil, jsonError);
            return;
        }
        
        NSString *jobId = json[@"job_id"];
        if (jobId) {
            completion(jobId, nil);
        } else {
            NSArray *messages = json[@"messages"];
            if (messages.count > 0) {
                NSString *errText = messages[0][@"err_text"];
                if (errText) {
                    NSError *messageError = [NSError errorWithDomain:errText code:-1 userInfo:nil];
                    completion(nil, messageError);
                    return;
                }
            }
            completion(nil, [NSError errorWithDomain:@"Unknown error" code:-1 userInfo:nil]);
        }
    }];
    
    [task resume];
}

- (void)checkBalanceWithCompletion:(void (^)(double, NSError * _Nullable))completion {
    NSString *endpoint = [self.baseURL stringByAppendingString:@"account/area/me/packages"];
    NSURL *url = [NSURL URLWithString:endpoint];
    
    if (!url) {
        NSError *error = [NSError errorWithDomain:@"Invalid URL" code:-1 userInfo:nil];
        completion(0, error);
        return;
    }
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"GET";
    [request setValue:@"application/json" forHTTPHeaderField:@"Accept"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request setValue:[self authHeader] forHTTPHeaderField:@"Authorization"];
    
    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            completion(0, error);
            return;
        }
        
        if (!data) {
            NSError *noDataError = [NSError errorWithDomain:@"No data received" code:-1 userInfo:nil];
            completion(0, noDataError);
            return;
        }
        
        NSError *jsonError;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
        
        if (jsonError) {
            completion(0, jsonError);
            return;
        }
        
        NSNumber *balance = json[@"total_balance"];
        if (balance) {
            completion(balance.doubleValue, nil);
        } else {
            NSError *formatError = [NSError errorWithDomain:@"Invalid response format" code:-1 userInfo:nil];
            completion(0, formatError);
        }
    }];
    
    [task resume];
}

- (void)getSendersWithCompletion:(void (^)(NSArray<NSString *> * _Nullable, NSError * _Nullable))completion {
    NSString *endpoint = [self.baseURL stringByAppendingString:@"account/area/senders"];
    NSURL *url = [NSURL URLWithString:endpoint];
    
    if (!url) {
        NSError *error = [NSError errorWithDomain:@"Invalid URL" code:-1 userInfo:nil];
        completion(nil, error);
        return;
    }
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"GET";
    [request setValue:@"application/json" forHTTPHeaderField:@"Accept"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request setValue:[self authHeader] forHTTPHeaderField:@"Authorization"];
    
    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            completion(nil, error);
            return;
        }
        
        if (!data) {
            NSError *noDataError = [NSError errorWithDomain:@"No data received" code:-1 userInfo:nil];
            completion(nil, noDataError);
            return;
        }
        
        NSError *jsonError;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
        
        if (jsonError) {
            completion(nil, jsonError);
            return;
        }
        
        NSArray *items = json[@"items"];
        if (items) {
            NSMutableArray *senders = [NSMutableArray array];
            for (NSDictionary *item in items) {
                NSString *senderName = item[@"sender_name"];
                if (senderName) {
                    [senders addObject:senderName];
                }
            }
            completion(senders, nil);
        } else {
            NSError *formatError = [NSError errorWithDomain:@"Invalid response format" code:-1 userInfo:nil];
            completion(nil, formatError);
        }
    }];
    
    [task resume];
}

@end

File: swift/JawalySMSService.swift
import Foundation

class JawalySMSService {
    private let baseURL = "https://api-sms.4jawaly.com/api/v1/"
    private let appID: String
    private let appSecret: String
    private let session: URLSession
    
    init(appID: String, appSecret: String) {
        self.appID = appID
        self.appSecret = appSecret
        self.session = URLSession.shared
    }
    
    private var authHeader: String {
        let credentials = "\(appID):\(appSecret)"
        let encodedCredentials = Data(credentials.utf8).base64EncodedString()
        return "Basic \(encodedCredentials)"
    }
    
    func sendSMS(message: String, numbers: [String], sender: String, completion: @escaping (Result<String, Error>) -> Void) {
        let endpoint = baseURL + "account/area/sms/send"
        guard let url = URL(string: endpoint) else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1)))
            return
        }
        
        let messageData: [String: Any] = [
            "messages": [
                [
                    "text": message,
                    "numbers": numbers,
                    "sender": sender
                ]
            ]
        ]
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(authHeader, forHTTPHeaderField: "Authorization")
        request.httpBody = try? JSONSerialization.data(withJSONObject: messageData)
        
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "No data received", code: -1)))
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    if let jobId = json["job_id"] as? String {
                        completion(.success(jobId))
                    } else if let messages = json["messages"] as? [[String: Any]],
                              let firstMessage = messages.first,
                              let errText = firstMessage["err_text"] as? String {
                        completion(.failure(NSError(domain: errText, code: -1)))
                    }
                }
            } catch {
                completion(.failure(error))
            }
        }
        task.resume()
    }
    
    func checkBalance(completion: @escaping (Result<Double, Error>) -> Void) {
        let endpoint = baseURL + "account/area/me/packages"
        guard let url = URL(string: endpoint) else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1)))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(authHeader, forHTTPHeaderField: "Authorization")
        
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "No data received", code: -1)))
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let balance = json["total_balance"] as? Double {
                    completion(.success(balance))
                } else {
                    completion(.failure(NSError(domain: "Invalid response format", code: -1)))
                }
            } catch {
                completion(.failure(error))
            }
        }
        task.resume()
    }
    
    func getSenders(completion: @escaping (Result<[String], Error>) -> Void) {
        let endpoint = baseURL + "account/area/senders"
        guard let url = URL(string: endpoint) else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1)))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(authHeader, forHTTPHeaderField: "Authorization")
        
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "No data received", code: -1)))
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let items = json["items"] as? [[String: Any]] {
                    let senders = items.compactMap { $0["sender_name"] as? String }
                    completion(.success(senders))
                } else {
                    completion(.failure(NSError(domain: "Invalid response format", code: -1)))
                }
            } catch {
                completion(.failure(error))
            }
        }
        task.resume()
    }
}
